src/os/bluestore/AvlAllocator.cc

```diff
@@ -240,13 +247,27 @@ int AvlAllocator::_allocate(
 
   const int free_pct = num_free * 100 / device_size;
   uint64_t start = 0;
-  /*
-   * If we're running low on space switch to using the size
-   * sorted AVL tree (best-fit).
-   */
+  // If we're running low on space, find a range by size by looking up in the size
+  // sorted tree (best-fit), instead of searching in the area pointed by cursor
   if (force_range_size_alloc ||
       max_size < range_size_alloc_threshold ||
       free_pct < range_size_alloc_free_pct) {
+    start = -1ULL;
+  } else {
+    /*
+     * Find the largest power of 2 block size that evenly divides the
+     * requested size. This is used to try to allocate blocks with similar
+     * alignment from the same area (i.e. same cursor bucket) but it does
+     * not guarantee that other allocations sizes may exist in the same
+     * region.
+     */
+    uint64_t align = size & -size;
+    ceph_assert(align != 0);
+    uint64_t* cursor = &lbas[cbits(align) - 1];
+    start = _pick_block_after(cursor, size, unit);
+    dout(20) << __func__ << " first fit=" << start << " size=" << size << dendl;
+  }
+  if (start == -1ULL) {
```

Comment:
With this modification we gave up first-fit mode for all follow-up attempts with reduced sizes when original request size hasn't been satisfied.  Is this intentional? Do you think we shouldn't try first fit mode any more when attempting to allocate with reduced size? I'm pretty uncertain myself on whether this is correct or not though.

Reply:
@ifed01 no, it's not intentional. i need to ponder over this a little bit.

Reply:
> With this modification we gave up first-fit mode for all follow-up attempts with reduced sizes when original request size hasn't been satisfied.

@ifed01 this is not true. for instance, assuming

- we need to offer 67KB chunk, and 
- the `unit` is 1KB,
- the maximum contiguous chunk in system is 64KB, so `max_size` is 64KB
- the maximum contiguous chunk in the area pointed by cursor for 64K aligned allocations is 32 KB.
- the maximum contiguous chunk in the area pointed by cursor for 32K aligned allocations is 32 KB.
- the maximum contiguous chunk in the area pointed by cursor for 1K aligned allocations is 5 KB.
- the total free space is 2MB. so the free percentage is higher than `range_size_alloc_free_pct`
- range_size_alloc_threshold is 1KB. as we always want to use first-fit if possible.

where the "area" is limited by `bluestore_avl_alloc_ff_max_search_count` and `bluestore_avl_alloc_ff_max_search_bytes` settings.

so we have following call sequence

1. _allocate(67KB) // greater than `max_size`
   1. best_fit returns 64 KB
2. _allocate(3KB) // less than `max_size`
   1. first_fit after 1KB cursor returns 3KB
3. fin


but i agree that, there are quite a few cases, we could have allocated chunks in the first_fit mode if we allow it to allocate smaller 
chunks. if we use the algorithm proposed by this PR, we have

1. _allocate(128KB) // greater than `max_size`
   1. best_fit returns 64 KB
2. _allocate(64KB) // less-or-equal than `max_size`
   1. first_fit after 1KB cursor fails
   2. best_fit returns 64 KB
3. fin

but if we *could* allow first-fit to try harder by allocating smaller chunks:

1. _allocate(128KB) // greater than `max_size`
   1. best_fit returns 64 KB
2. _allocate(64KB) // less-or-equal than `max_size`
    1. first_fit after 1KB cursor fails, but it managed to allocate 32KB in 32KB aligned area.
3. _allocate(32KB)
    1. first_fit after 32KB cursor returns 32KB. because it is allowed to move further in a new search. but this chunk is *not* located next to the previous 32KB chunk, otherwise the previous first_fit call would have returned 64KB.

so, 

1. this allocator does allow switch back to first-fit mode even if the previous call ends up with a reduced size chunk
2. allowing first-fit to try harder by allocating smaller chunk if it is not able to fulfill the required size does not help with the clumping the allocating the adjacent allocation request or the sub-requests belonging to the same original request.

Reply:
@tchaikov - not sure I got all the points above but it looks like we're talking about different scenarios.
The one of mine looks like the following (based on the recent case I made a fix for in #41369):
1) _allocate(4M) is invoked
2) max_size = 4M and there is the only 4M chunk in the pool, it's unaligned but plenty of e.g. 2MB chunks are available. Hence first_fit mode is chosen at first step but _block_picker_after() call returns -1 due to 4M chunk misalignment.
So far there is no logic difference between your patch and the preceding implementation (#41369) .
3) Then requested size is dropped to 2MB and _allocate retries with 2MB chunk size. 
Prior implementation did that in first-fit mode too which provided geo locality(via cursor usage) for 2MB allocations.
With your patch this reattempt is going to be in best-fit mode unconditionally which looks like a sort of regression. 
Again not sure it's that important...

Reply:
@ifed01

1. _allocate(4M) is invoked
2. first_fit mode is chosen at first step but _block_picker_after() call returns -1
3. "Then requested size is dropped to 2MB"

i don't think the requested size is dropped here. in my change, if first_fit mode fails, `_allocate(4M)` just switches over to best_fit mode without changing the requested size. and in your case, since we have non-aligned 4M chunks in the pool, best_fit just returns the first found 4M chunk to the caller.